File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/integration/booking/adapter/BookingResult.java
package it.giannibombelli.wsc2026.giftcard.application.integration.booking.adapter;

import it.giannibombelli.wsc2026.booking.integration.giftcard.BookingResultIntegrationEvent;
import it.giannibombelli.wsc2026.common.utils.Require;
import it.giannibombelli.wsc2026.giftcard.application.commands.CreditGiftCard;
import it.giannibombelli.wsc2026.giftcard.application.commands.RefundGiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;

/**
 * Anti-Corruption Layer che traduce la Published Language di {@code booking} nei command
 * interni di {@code giftcard}: l'unico contratto condiviso è {@link BookingResultIntegrationEvent}.
 */
public final class BookingResult {

    public CreditGiftCard adaptCredit(BookingResultIntegrationEvent event) {
        Require.requireArgument(event, "Booking result integration event");
        return switch (event) {
            case BookingResultIntegrationEvent.BookingCompletedIntegrationEvent it ->
                new CreditGiftCard(new GiftCardId(it.giftCardReference()), it.amount());
            case BookingResultIntegrationEvent.BookingRefusedIntegrationEvent it ->
                new CreditGiftCard(new GiftCardId(it.giftCardReference()), it.amount());
            case BookingResultIntegrationEvent.BookingRejectedIntegrationEvent ignored -> null;
        };
    }

    public RefundGiftCard adaptRefund(BookingResultIntegrationEvent.BookingRejectedIntegrationEvent event) {
        Require.requireArgument(event, "Booking rejected integration event");
        return new RefundGiftCard(new GiftCardId(event.giftCardReference()), event.amount());
    }
}

File: typescript/src/giftcard/application/integration/booking/adapter/bookingResult.ts
// Anti-Corruption Layer: isola `giftcard` dalle evoluzioni del modello interno di `booking`;
// l'unico contratto condiviso è la Published Language `BookingResultIntegrationEvent`.

import type {
  BookingRejectedIntegrationEvent,
  BookingResultIntegrationEvent,
} from '@/booking/integration/giftcard/bookingResultIntegrationEvent.js';
import type { CreditGiftCard } from '@/giftcard/application/commands/creditGiftCard.js';
import type { RefundGiftCard } from '@/giftcard/application/commands/refundGiftCard.js';
import { creditGiftCard } from '@/giftcard/application/commands/creditGiftCard.js';
import { refundGiftCard } from '@/giftcard/application/commands/refundGiftCard.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export class BookingResult {
  adaptCredit(event: BookingResultIntegrationEvent): CreditGiftCard | null {
    requireArgument(event, 'Booking result integration event');

    switch (event.kind) {
      case 'BookingCompletedIntegrationEvent':
      case 'BookingRefusedIntegrationEvent':
        return creditGiftCard(new GiftCardId(Uuid.fromString(event.giftCardReference)), event.amount);
      case 'BookingRejectedIntegrationEvent':
        return null;
    }
  }

  adaptRefund(event: BookingRejectedIntegrationEvent): RefundGiftCard {
    requireArgument(event, 'Booking rejected integration event');
    return refundGiftCard(new GiftCardId(Uuid.fromString(event.giftCardReference)), event.amount);
  }
}
